qcengine/programs/madness/harvester.py

```diff
@@ -0,0 +1,225 @@
+import re
+import json
+import logging
+from decimal import Decimal
+from typing import Tuple
+
+import numpy as np
+import qcelemental as qcel
+from qcelemental.models import Molecule
+from qcelemental.models.results import AtomicResultProperties
+from qcelemental.molparse import regex
+
+from ..util import PreservingDict
+
+logger = logging.getLogger(__name__)
+
+
+def harvest_output(outtext: str) -> Tuple[PreservingDict, Molecule, list, str, str]:
+    """Function to read an entire MADNESS output file.
+
+    Reads all of the different "line search" segments of a file and returns
+    values from the last segment for which a geometry was written.
+
+    Args:
+        outtext (str): Output written to stdout
+    Returns:
+        - (PreservingDict) Variables extracted from the output file in the last complete step
+        - (Molecule): Molecule from the last complete step
+        - (list): Gradient from the last complete step
+        - (str): Version string
+        - (str): Error message, if any
+    """
+
+    # Loop over all steps
+    pass_psivar = []
+    pass_coord = []
+    pass_grad = []
+    for outpass in re.split(r"Converged!", outtext, re.MULTILINE):
+        psivar, madcoord, madgrad, version, error = harvest_outfile_pass(outpass)
+        pass_psivar.append(psivar)## all the variables extracted
+        pass_coord.append(madcoord)
+        pass_grad.append(madgrad)
+
+    # Determine which segment contained the last geometry
+    retindx = -1 #if pass_coord[-1] else -2
+    return pass_psivar[retindx], pass_coord[retindx], pass_grad[retindx], version, error
+
+
+def harvest_outfile_pass(outtext):
+    """Function to read Madness output file *outtext* and parse important
+    quantum chemical information from it in
+
+    """
+    psivar = PreservingDict()
+    psivar_coord = None
+    psivar_grad = None
+    version = ""
+    error = ""  # TODO (wardlt): The error string is never used.
+
+    NUMBER = r"(?x:" + regex.NUMBER + ")"
+    # fmt: off
```

Comment:
I just fixed this in nwchem. it's better to have fmt: off/on multiple times within regexes rather than suppressing it for a whole file.